corgi/tasks/prod_defs.py

```diff
@@ -67,6 +69,89 @@ def update_products() -> None:
+    cpe_version: str,
+    cpe: str,
+    el_by_cpe: defaultdict[str, list],
+    cpes_by_stream: defaultdict[str, list],
+):
+    """Iterate product stream versions ordered by length descending, looking for exact matches to
+    the stream's parent ProductVersion cpe. If a match is found, store the match along with any
+    el suffixes."""
+    for version_stream in version_streams:
+        if version_stream.version.removesuffix(".z") == cpe_version:
+            if cpe in el_by_cpe:
+                for el in el_by_cpe[cpe]:
+                    cpes_by_stream[version_stream.name].append(cpe + "::" + el)
+            else:
+                cpes_by_stream[version_stream.name].append(cpe)
+
+
+def _split_cpe_version(cpe_part: str, versions_by_cpe: defaultdict[str, set]):
+    version_split = cpe_part.rsplit(":", 1)
+    cpe_name = version_split[0]
+    cpe_version = version_split[-1]
+    versions_by_cpe[cpe_name].add(cpe_version)
+
+
+def _split_cpe(cpe: str, versions_by_cpe: defaultdict[str, set], el_by_cpe: defaultdict[str, list]):
+    """Splits a cpe such as 'cpe:/a:redhat:openshift_ironic:4.13:el9' into
+    'cpe:/a:redhat:openshift_ironic', '4.13', and 'el9'
+    It builds 2 dictionaries, one with a set of versions grouped by cpe_name and another
+    with el suffixes grouped by 'cpe_name + version'.
+    We do this so that we can iterate versions looking for matches without el suffixes. Once
+    we find a match, we add back the suffixes"""
+    variant_split = cpe.rsplit("::")
+    if len(variant_split) == 2:
+        el_part = variant_split[1]
+        cpe_part = variant_split[0]
+        el_by_cpe[cpe_part].append(el_part)
+        _split_cpe_version(cpe_part, versions_by_cpe)
+    elif len(variant_split) == 1:
+        _split_cpe_version(cpe, versions_by_cpe)
+    else:
+        raise ValueError(f"More than one el qualifier '::' in {cpe}")
+
+
+def match_stream_cpes(product_version: ProductVersion) -> None:
+    """Given a Product Version with Product Stream foreign keys and cpes_matching_patterns
+    This will match those cpes to the stream children using the stream versions"""
+    versions_by_cpe: defaultdict[str, set] = defaultdict(set)
+    el_by_cpe: defaultdict[str, list] = defaultdict(list)
+    for cpe in product_version.cpes_matching_patterns:
+        _split_cpe(cpe, versions_by_cpe, el_by_cpe)
+
+    cpes_by_stream: defaultdict[str, list] = defaultdict(list)
+    version_streams = product_version.productstreams.all()
+    for cpe_name, versions in versions_by_cpe.items():
+        for cpe_version in versions:
+            cpe = cpe_name + ":" + cpe_version
+            _match_stream_version_to_cpe_version(
+                version_streams, cpe_version, cpe, el_by_cpe, cpes_by_stream
+            )
+
+    for stream, cpes in cpes_by_stream.items():
+        product_stream = ProductStream.objects.get(name=stream)
+        product_stream.cpes_matching_patterns = cpes
+        product_stream.save()
```

Comment:
I think the code that calls this function is already running in a transaction, but let's use an inner transaction or .update() here in case we ever start calling this function from somewhere else. .update() is also a little faster than calling .get() + making a change + calling .save():
```suggestion
    for stream, cpes in cpes_by_stream.items():
        ProductStream.objects.filter(name=stream).update(cpes_matching_patterns=cpes)
```
Or if we never expect to call this code anywhere else, let's add an _ underscore in front of the method name.

Reply:
Good idea, I'll use .update() function to make it atomic.  I'll also add an underscore to the method name.